File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/data/api/deserializers/ApiResponseDeserializer.java
package com.crpr.androidcinema.data.api.deserializers;

import com.crpr.androidcinema.data.api.responses.ApiResponse;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class ApiResponseDeserializer<T> implements JsonDeserializer<ApiResponse<T>> {

    @Override
    public ApiResponse<T> deserialize(JsonElement je, Type typeOfT,
                                         JsonDeserializationContext context) throws JsonParseException {

        ApiResponse<T> response = new ApiResponse<>();
        response.setPage(je.getAsJsonObject().get("page").getAsInt());
        response.setPage(je.getAsJsonObject().get("total_pages").getAsInt());
        response.setPage(je.getAsJsonObject().get("total_results").getAsInt());

        JsonArray results = je.getAsJsonObject().get("results").getAsJsonArray();
        List<T> rvalueList = new ArrayList<>(results.getAsJsonArray().size());

        for(JsonElement element : results){
            rvalueList.add(new Gson().fromJson(element, ((ParameterizedType) typeOfT).getActualTypeArguments()[0]));
        }

        response.setResults(rvalueList);

        return response;
    }
}

File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/data/api/deserializers/ApiResponseDeserializer.kt
package com.crpr.androidcinema.data.api.deserializers

import com.crpr.androidcinema.data.api.responses.ApiResponse
import com.google.gson.*
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.*

class ApiResponseDeserializer<T> : JsonDeserializer<ApiResponse<T>> {

    @Throws(JsonParseException::class)
    override fun deserialize(je: JsonElement, typeOfT: Type,
                             context: JsonDeserializationContext): ApiResponse<T> {

        val response = ApiResponse<T>()
        response.page = je.asJsonObject.get("page").asInt
        response.totalPages = je.asJsonObject.get("total_pages").asInt
        response.totalResults = je.asJsonObject.get("total_results").asInt

        val results = je.asJsonObject.get("results").asJsonArray
        val rvalueList = ArrayList<T>(results.asJsonArray.size())

        for (element in results) {
            rvalueList.add(Gson().fromJson<T>(element, (typeOfT as ParameterizedType).actualTypeArguments[0]))
        }

        response.results = rvalueList

        return response
    }
}